_posts/2023-10-13-authz-keto-introduction.md

```diff
@@ -0,0 +1,273 @@
+* **Intuitive Nature**: Keto is designed to be self-explanatory, minimizing the learning curve and ensuring that users find the system approachable.
+* **Familiarity**: By adhering to familiar paradigms and practices, Keto reduces the onboarding time for developers and system administrators.
+* **Ease of Modification**: Any changes or updates to the authorization configurations can be carried out with high confidence, ensuring system stability.
+* **Editorial Support**: Thanks to the Ory Permission Language, which is a subset of Typescript, Keto is compatible with numerous editors. This ensures users have the flexibility and familiarity in their choice of tools.
+
+## Core Concepts of Ory Keto
+
+1. **Adoption of the Relation Tuple Model**: Ory Keto adapts the Zanzibar-inspired relation tuples model. As we’ve previously elaborated, a relation tuple succinctly describes a relationship in a format like `namespace:object#relation@subject`. Through these tuples, Ory Keto can express and evaluate intricate hierarchical relationships and permissions across a multitude of objects and users, ensuring a granular and fine-tuned access control mechanism.
+2. **Ory Permission Language (OPL)**: To complement its adoption of relation tuples, Ory Keto introduces the Ory Permission Language (OPL). Conceptualized as a subset of TypeScript, OPL facilitates the definition and evaluation of policies and permissions within the Keto environment. By leveraging a familiar syntax style and structure, OPL provides developers with a more intuitive and streamlined process to articulate complex authorization policies, while maintaining efficiency and precision.
+
+### Ory Keto Namespaces
+
+In Ory Keto, namespaces are essentially the building blocks that aid in grouping related objects. To define these namespaces, Ory Keto uses TypeScript's class structures. 
+
+```
+import { Namespace, Context } from "@ory/keto-namespace-types"
+
+class User implements Namespace {}
+class Document implements Namespace {}
+class Folder implements Namespace {}
+```
+
+### Ory Keto Relations
+
+In Ory Keto, defining relations is at the heart of the model. Let's illustrate this with a hands-on example featuring two namespaces: Document and User.
+Within the Document namespace, a user is designated as either an owner, a viewer, or an editor of a document through specific relations.
+
+The core of authorization revolves around permissions, such as 'view' or 'edit', rather than checking relations like 'owner' or 'editor'. The concrete permission is checked against the relationships. Within the Ory Permission Language, permissions are declared as functions inside the permits property of the corresponding namespace. Let's explore how permissions might be outlined for our Document namespace:
+
+```
+class Document implements Namespace {
+  related: {
```

Comment:
up of point, the flow of the text was excellent. it was a fascinating story about the Google paper. however, I personally struggled at this point. 

I am confused because we learnt about the relation tuples earlier, but this code seems like it has nothing to do with that. why are we not using tuples here? what is the relationship between the tuples and this code? is this code for a specific tuple? what is the purpose of this code? maybe the purpose is to define which relation tuples are legal?

Reply:
I thought we would insert relations into a database and it would perform a graph theory algorithm to determine whether a connection across multiple relationtuples exists, that connects the user and the document. however this looks like some good simple old boolean permission checks

Reply:
The OPL incorporates both the model (namespaces), and the rules which evaluate permissions by navigating through relations. However, OPL itself does not perform the actual relation traversal. Instead, this is handled internally by Keto. While Keto doesn't directly utilize TypeScript, it operates purely on Go. The OPL is parsed by Keto during its initialization.

This approach is somewhat different from the original modeling presented in the Google Zanzibar paper. That modeling was expressed in a more scientific manner, using mathematical notations and representations. The audience wasn't the everyday developers but more advanced researchers and professionals. 

The motivation to use TypeScript (TS) for OPL was to make it more developer-friendly. TypeScript is a widely-recognized language with extensive editor support, making it a more accessible and familiar tool for many developers. This adoption aimed to simplify the tooling and make it easier for the majority of developers to understand and work with.

Reply:
I suggest that you tell the reader the goal of the code before you show the code. for example:

"the following code defines a simple permission pattern, which keto then uses to build relation tuples internally." or something similar :)

Reply:
I adapted it a bit and added a chapter that handles the creation of relations explicitly.

Reply:
https://github.com/sw-code/sw-code.github.io/pull/5

Reply:
much better in my opinion, thanks:)